### PR TITLE
QPT-11: Implement a Logger in Testing

### PR DESCRIPTION
**Pull request by aaron-humerickhouse**
Add logging for tests and actions to a file
Keeps logs in the console as well

Proof:
From file `1575414593343-Bigcommerce_page_should_have_the_right_title.log`:

```
[2019-12-03T17:09:53.348] [INFO] test - --- Starting test Bigcommerce page should have the right title ---
[2019-12-03T17:09:53.350] [INFO] test - Navigating to bigcommerce.com
[2019-12-03T17:09:57.229] [INFO] test - --- Ending test Bigcommerce page should have the right title ---
```

From console:

```
[0-0] --- Starting test Bigcommerce page should have the right title ---
[0-0] Navigating to bigcommerce.com
[0-0] --- Ending test Bigcommerce page should have the right title ---
```

### CODE REVIEW

**Comment by lorentzkim** [comment]
:beers: Was wondering, on a general level, were there no other functionality we could've leveraged to programmatically (through the test code themselves) to describe what's happening to log them, rather than `console.log`? Or is that not available in whatever we're using for big-design?

**Comment by KenZSyd** [comment]
> 🍻 Was wondering, on a general level, were there no other functionality we could've leveraged to programmatically (through the test code themselves) to describe what's happening to log them, rather than `console.log`? Or is that not available in whatever we're using for big-design?

You mean the codeception style?

```
✔ SharedWishlistCest: Create public wishlist (13.87s)
✔ SharedWishlistCest: My friend can open my shared wishlist (3.98s)
```

**Comment by lorentzkim** [comment]
> 
> You mean the Codeception style (writing code without calling logging functions but on console we see something)?
> 
> ```
> ✔ SharedWishlistCest: Create public wishlist (13.87s)
> ✔ SharedWishlistCest: My friend can open my shared wishlist (3.98s)
> ```

Something like it;

```
await browser.url('https://www.bigcommerce.com/')
```

Wondering if code like that can automatically be converted to info log entries, and displayable by debug level etc.

File: packages/tests/wdio.conf.js [hunk 1]
```diff
@@ -221,8 +231,39 @@ exports.config = {
...
+        let filename = test['fullName'].replace(/\s/gi, '_');
+        let file = `${logsPath}/${test['start']}-${filename}.log`;
+
+        const logger = Logger.new(file);
+
+        console = (function(originalConsole){
+            return {
+                log: function(text){
+                    originalConsole.log(text);
+                    logger.info(text)
```
Comment (lorentzkim) [inline packages/tests/wdio.conf.js]: It's not `logger.log` here?

Comment (aaron-humerickhouse) [reply packages/tests/wdio.conf.js]: log4js doesn't have `.log`

console.log is an info message
https://developers.google.com/web/tools/chrome-devtools/console/log
> Each console method is assigned a severity level: Verbose, Info, Warning, or Error. For example, console.log() is an Info-level message

**Comment by aaron-humerickhouse** [comment]
wdio does have logging like that, but it's driven down into the actual webdriver logs:

```
2019-12-05T15:08:50.254Z INFO @wdio/local-runner: Run worker command: run
2019-12-05T15:08:51.272Z INFO webdriverio: Initiate new session using the webdriver protocol
2019-12-05T15:08:51.275Z INFO webdriver: [POST] http://127.0.0.1:4444/wd/hub/session
2019-12-05T15:08:51.275Z INFO webdriver: DATA { capabilities:
   { alwaysMatch: { browserName: 'firefox' }, firstMatch: [ {} ] },
  desiredCapabilities: { browserName: 'firefox' } }
2019-12-05T15:09:04.424Z ERROR @wdio/utils:shim: Error: Bad response code: Expected (200), Received (401)!
    at Request._request.default.get [as _callback] (/Users/aaron.humerickhouse/dev/big-design/node_modules/@wdio/browserstack-service/build/service.js:122:23)
    at Request.self.callback (/Users/aaron.humerickhouse/dev/big-design/node_modules/request/request.js:185:22)
    at Request.emit (events.js:189:13)
    at Request.EventEmitter.emit (domain.js:441:20)
    at Request.<anonymous> (/Users/aaron.humerickhouse/dev/big-design/node_modules/request/request.js:1161:10)
    at Request.emit (events.js:189:13)
    at Request.EventEmitter.emit (domain.js:441:20)
    at IncomingMessage.<anonymous> (/Users/aaron.humerickhouse/dev/big-design/node_modules/request/request.js:1083:12)
    at Object.onceWrapper (events.js:277:13)
    at IncomingMessage.emit (events.js:194:15)
2019-12-05T15:09:04.455Z INFO webdriver: COMMAND navigateTo("https://www.bigcommerce.com/")
2019-12-05T15:09:04.456Z INFO webdriver: [POST] http://127.0.0.1:4444/wd/hub/session/a539d3ba-5963-ba48-929c-8ab9f700d069/url
2019-12-05T15:09:04.456Z INFO webdriver: DATA { url: 'https://www.bigcommerce.com/' }
2019-12-05T15:09:06.351Z INFO webdriver: COMMAND getTitle()
2019-12-05T15:09:06.351Z INFO webdriver: [GET] http://127.0.0.1:4444/wd/hub/session/a539d3ba-5963-ba48-929c-8ab9f700d069/title
2019-12-05T15:09:06.357Z INFO webdriver: RESULT The Future of Commerce Is Yours | BigCommerce
2019-12-05T15:09:06.603Z INFO webdriver: COMMAND deleteSession()
2019-12-05T15:09:06.603Z INFO webdriver: [DELETE] http://127.0.0.1:4444/wd/hub/session/a539d3ba-5963-ba48-929c-8ab9f700d069
```

It's incredibly messy when the tests get large.  Taking individual steps and making them logs, like codeception does, is not common.  Tests aren't cleanly written generally.

What we will do, is add the logging to the actions instead of the test.  So when we have a method checkbox.select(#selector), we can have the message logged within that method stating something like `Selecting checkbox with selector: #selector`.

Remember, this is POC, let's not get caught up over some of these details yet.  It will be small and we can address some things later under the advisement of JS experts if we continue.

**Comment by lorentzkim** [comment]
All good, was curious.

**Comment by KenZSyd** [comment]
@aaron-humerickhouse I guess its better to invite front-end devs to review this.

File: packages/tests/functional/basic.ts [hunk 2]
```diff
@@ -1,7 +1,8 @@
 describe('Bigcommerce page', () => {
     it('should have the right title', async () => {
+        console.log('Navigating to bigcommerce.com')
```
Comment (Sure-Round) [inline packages/tests/functional/basic.ts]: Do we need the logging?

File: packages/tests/functional/basic.ts [hunk 3]
```diff
@@ -1,7 +1,8 @@
...
         await browser.url('https://www.bigcommerce.com/')
         const title = await browser.getTitle()
         expect(title).toEqual('The Future of Commerce Is Yours | BigCommerce')
     })
-})
\ No newline at end of file
+})
```
Comment (rtalvarez) [inline packages/tests/functional/basic.ts]: Sorry I'm a little late to the party, but let's consider linting these files with the same configs to keep the code consistent with the rest of the codebase

File: packages/tests/wdio.conf.js [hunk 4]
```diff
@@ -1,3 +1,13 @@
+const Logger = require('./logger.js');
+
+const testPath = `.`;
```
Comment (rtalvarez) [inline packages/tests/wdio.conf.js]: btw no need to use backticks ` when not interpolating values, a standard single quote will sufice

File: packages/tests/wdio.conf.js [hunk 5]
```diff
@@ -221,8 +231,39 @@ exports.config = {
...
+    beforeTest: function (test, context) {
+        // Override console to allow for STDOUT and file logging with one command
+        originalConsole = console
+
+        let filename = test['fullName'].replace(/\s/gi, '_');
+        let file = `${logsPath}/${test['start']}-${filename}.log`;
+
+        const logger = Logger.new(file);
+
+        console = (function(originalConsole){
```
Comment (rtalvarez) [inline packages/tests/wdio.conf.js]: again mostly a stylistic comment that could be picked up by a linter, but we normally prefer arrow functions over function declarations

File: packages/tests/functional/basic.ts (hunk 2 above)
Comment (aaron-humerickhouse) [reply packages/tests/functional/basic.ts]: This was to make sure it worked, it'll be removed from the hello-world example

File: packages/tests/functional/basic.ts (hunk 3 above)
Comment (aaron-humerickhouse) [reply packages/tests/functional/basic.ts]: :+1:

**Pull request merged**